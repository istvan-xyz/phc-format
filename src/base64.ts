// Lightweight base64 helpers that work in Node, Deno, and browsers without requiring Buffer types.
// Encoding omits padding (=) to match PHC formatting behavior used in this project.

function bytesToBinaryString(bytes: Uint8Array): string {
    let result = '';
    for (const b of bytes) {
        result += String.fromCharCode(b);
    }
    return result;
}

function binaryStringToBytes(binary: string): Uint8Array {
    const out = new Uint8Array(binary.length);
    for (let i = 0; i < binary.length; i++) {
        out[i] = binary.charCodeAt(i);
    }
    return out;
}

export function bytesToBase64(bytes: Uint8Array): string {
    // Prefer global btoa when available (Deno, browsers)
    let b64: string;
    const g = globalThis as unknown as {
        btoa?: (data: string) => string;
        atob?: (data: string) => string;
        // Intentionally loose shape to avoid Node type dependency
        Buffer?: unknown;
    };
    if (typeof g.btoa === 'function') {
        b64 = g.btoa(bytesToBinaryString(bytes));
    } else if (typeof g.Buffer !== 'undefined') {
        // Node fallback
        // Use node's Buffer without importing types
        const BufferLike = g.Buffer as unknown as {
            from: (input: unknown) => { toString: (enc: string) => string };
        };
        const nodeBuf = BufferLike.from(bytes);
        b64 = nodeBuf.toString('base64');
    } else {
        // Very old environments; last resort using TextEncoder/Decoder
        // Note: This will not work without a polyfill for btoa/atob
        throw new Error('No base64 encoder available in this environment');
    }
    // Strip trailing padding as PHC examples avoid '=' chars
    return b64.replace(/=+$/u, '');
}

export function base64ToBytes(b64: string): Uint8Array {
    // Re-add padding to make length a multiple of 4
    const padLength = (4 - (b64.length % 4)) % 4;
    const padded = b64 + '='.repeat(padLength);

    const g = globalThis as unknown as {
        atob?: (data: string) => string;
        Buffer?: unknown;
    };
    if (typeof g.atob === 'function') {
        const binary = g.atob(padded);
        return binaryStringToBytes(binary);
    }
    if (typeof g.Buffer !== 'undefined') {
        const BufferLike = g.Buffer as unknown as {
            from: (input: unknown, enc: string) => Uint8Array;
        };
        const buf = BufferLike.from(padded, 'base64');
        return new Uint8Array(buf);
    }
    throw new Error('No base64 decoder available in this environment');
}

// In Node, prefer returning Buffer instances for compatibility with existing code/tests.
export function asRuntimeBytes(bytes: Uint8Array): Uint8Array {
    const g = globalThis as unknown as { Buffer?: unknown };
    if (typeof g.Buffer !== 'undefined') {
        const BufferLike = g.Buffer as unknown as {
            from: (input: unknown) => Uint8Array;
        };
        return BufferLike.from(bytes);
    }
    return bytes;
}
