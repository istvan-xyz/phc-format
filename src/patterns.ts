export const idRegex = /^[a-z0-9-]{1,32}$/;
export const nameRegex = /^[a-z0-9-]{1,32}$/;
export const valueRegex = /^[a-zA-Z0-9/+.-]+$/;
