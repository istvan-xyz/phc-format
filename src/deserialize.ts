import { asRuntimeBytes, base64ToBytes } from './base64';
import { idRegex, nameRegex, valueRegex } from './patterns';

const b64Regex = /^([a-zA-Z0-9/+.-]+|)$/;
const decimalRegex = /^((-)?[1-9]\d*|0)$/;
const versionRegex = /^v=(\d+)$/;

const keyValueStringToObject = (string: string): { [key: string]: unknown } => {
    const object: { [key: string]: string } = {};

    string.split(',').forEach(ps => {
        const tokens = ps.split('=');
        if (tokens.length < 2) {
            throw new TypeError('params must be in the format name=value');
        }

        object[tokens.shift() as string] = tokens.join('=');
    });

    return object;
};

/**
 * Parses data from a PHC string.
 * @param  {string} phcString A PHC string to parse.
 * @return {Object} The object containing the data parsed from the PHC string.
 */
export default function deserialize(phcString: string) {
    if (phcString === '') {
        throw new TypeError('phcString must be a non-empty string');
    }

    if (phcString[0] !== '$') {
        throw new TypeError('phcString must contain a $ as first char');
    }

    const fields = phcString.split('$');

    // Remove first empty $
    fields.shift();

    // Parse Fields
    let maxFields = 5;

    if (!versionRegex.test(fields[1])) maxFields--;
    if (fields.length > maxFields) {
        throw new TypeError(`phcString contains too many fields: ${fields.length}/${maxFields}`);
    }

    // Parse Identifier
    const id = fields.shift();

    if (!id) {
        throw new Error('id cannot be undefined at this point.');
    }

    if (!idRegex.test(id)) {
        throw new TypeError(`id must satisfy ${idRegex}`);
    }

    let version;

    // Parse Version
    if (versionRegex.test(fields[0])) {
        const versionString = fields.shift();

        if (!versionString) {
            throw new Error('paramString cannot be undefined at this point.');
        }

        version = parseInt((versionString.match(versionRegex) as RegExpMatchArray)[1], 10);
    }

    let hash: Uint8Array | undefined;
    let salt: Uint8Array | undefined;
    if (b64Regex.test(fields[fields.length - 1])) {
        if (fields.length > 1 && b64Regex.test(fields[fields.length - 2])) {
            // Parse Hash
            hash = asRuntimeBytes(base64ToBytes(fields.pop() as string));
            // Parse Salt
            salt = asRuntimeBytes(base64ToBytes(fields.pop() as string));
        } else {
            // Parse Salt
            salt = asRuntimeBytes(base64ToBytes(fields.pop() as string));
        }
    }

    // Parse Parameters
    let params: { [key: string]: unknown } | undefined;

    if (fields.length > 0) {
        const paramString = fields.pop();

        if (!paramString) {
            throw new Error('paramString cannot be undefined at this point.');
        }

        const currentParams = keyValueStringToObject(paramString);

        if (!Object.keys(currentParams).every(p => nameRegex.test(p))) {
            throw new TypeError(`params names must satisfy ${nameRegex}`);
        }

        const pv = Object.values(currentParams);
        if (!pv.every(v => valueRegex.test(v as string))) {
            throw new TypeError(`params values must satisfy ${valueRegex}`);
        }

        const pk = Object.keys(currentParams);
        // Convert Decimal Strings into Numbers
        pk.forEach(k => {
            currentParams[k] = decimalRegex.test(currentParams[k] as string)
                ? parseInt(currentParams[k] as string, 10)
                : currentParams[k];
        });

        params = currentParams;
    }

    if (fields.length > 0) {
        throw new TypeError(`phcString contains unrecognized fields: ${fields}`);
    }

    // Build the output object
    const result: {
        id?: string;
        version?: number;
        params?: { [key: string]: unknown };
        salt?: Uint8Array;
        hash?: Uint8Array;
    } = { id };

    if (version) {
        result.version = version;
    }

    if (params) {
        result.params = params;
    }

    if (salt) {
        result.salt = salt;
    }

    if (hash) {
        result.hash = hash;
    }

    return result;
}
