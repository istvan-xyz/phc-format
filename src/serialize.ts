// cspell:words phcobj, phcstr, Valto, strpar, pchstr, maxf, parstr
import { idRegex, nameRegex, valueRegex } from './patterns';

function objectToKeyValueString(object: { [key: string]: unknown }) {
    return Object.entries(object)
        .map(([key, value]) => [key, value].join('='))
        .join(',');
}

/**
 * Generates a PHC string using the data provided.
 * @param  {Object} opts Object that holds the data needed to generate the PHC
 * string.
 * @param  {string} opts.id Symbolic name for the function.
 * @param  {Number} [opts.version] The version of the function.
 * @param  {Object} [opts.params] Parameters of the function.
 * @param  {Buffer} [opts.salt] The salt as a binary buffer.
 * @param  {Buffer} [opts.hash] The hash as a binary buffer.
 * @return {string} The hash string adhering to the PHC format.
 */
export default function serialize(opts: {
    id: string;
    version?: number;
    params?: { [key: string]: string | Buffer | number };
    salt?: Buffer;
    hash?: Buffer;
}) {
    const fields = [''];

    if (!idRegex.test(opts.id)) {
        throw new TypeError(`id must satisfy ${idRegex}`);
    }

    fields.push(opts.id);

    if (typeof opts.version !== 'undefined') {
        if (opts.version < 0 || !Number.isInteger(opts.version)) {
            throw new TypeError('version must be a positive integer number');
        }

        fields.push(`v=${opts.version}`);
    }

    // Parameter Validation
    const { params } = opts;
    if (typeof params !== 'undefined') {
        const pk = Object.keys(params);
        if (!pk.every(p => nameRegex.test(p))) {
            throw new TypeError(`params names must satisfy ${nameRegex}`);
        }

        // Convert Numbers into Numeric Strings and Buffers into B64 encoded strings.
        pk.forEach(k => {
            if (typeof params[k] === 'number') {
                params[k] = (params[k] as string).toString();
            } else if (Buffer.isBuffer(params[k])) {
                [params[k]] = (params[k] as Buffer).toString('base64').split('=');
            }
        });
        const pv = Object.values(params);
        if (!pv.every(v => typeof v === 'string')) {
            throw new TypeError('params values must be strings');
        }

        if (!pv.every(v => valueRegex.test(v as string))) {
            throw new TypeError(`params values must satisfy ${valueRegex}`);
        }

        const strpar = objectToKeyValueString(params);
        fields.push(strpar);
    }

    if (typeof opts.salt !== 'undefined') {
        fields.push(opts.salt.toString('base64').split('=')[0]);

        if (typeof opts.hash !== 'undefined') {
            // Hash Validation
            if (!Buffer.isBuffer(opts.hash)) {
                throw new TypeError('hash must be a Buffer');
            }

            fields.push(opts.hash.toString('base64').split('=')[0]);
        }
    }

    // Create the PHC formatted string
    const phcString = fields.join('$');

    return phcString;
}
