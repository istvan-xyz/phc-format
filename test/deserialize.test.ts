import { expect, test } from 'vitest';
import { deserialize } from '../src';
import { deserialized, serialized } from './fixtures';

test('should deserialize correct phc strings', () => {
    serialized.forEach((_, i) => {
        expect(deserialize(serialized[i])).toEqual(deserialized[i]);
    });
});

// cspell:disable
test('should thow errors if trying to deserialize an invalid phc string', () => {
    expect(() => {
        deserialize('');
    }).toThrowErrorMatchingSnapshot();

    expect(() => {
        deserialize('a$invalid');
    }).toThrowErrorMatchingSnapshot();

    expect(() => {
        deserialize('$b$c$d$e$f');
    }).toThrowErrorMatchingSnapshot();

    expect(() => {
        deserialize('invalid');
    }).toThrowErrorMatchingSnapshot();

    expect(() => {
        deserialize('$i_n_v_a_l_i_d');
    }).toThrowErrorMatchingSnapshot();

    expect(() => {
        deserialize('$pbkdf2$rounds_=1000');
    }).toThrowErrorMatchingSnapshot();

    expect(() => {
        deserialize('$pbkdf2$rounds=1000@');
    }).toThrowErrorMatchingSnapshot();

    expect(() => {
        deserialize('$pbkdf2$rounds:1000');
    }).toThrowErrorMatchingSnapshot();

    expect(() => {
        deserialize('$argon2i$unrecognized$m=120,t=5000,p=2$EkCWX6pSTqWruiR0');
    }).toThrowErrorMatchingSnapshot();

    expect(() => {
        deserialize('$argon2i$unrecognized$v=19$m=120,t=5000,p=2$EkCWX6pSTqWruiR0');
    }).toThrowErrorMatchingSnapshot();

    expect(() => {
        deserialize('$argon2i$v=19$unrecognized$m=120,t=5000,p=2$EkCWX6pSTqWruiR0');
    }).toThrowErrorMatchingSnapshot();
});
// cspell:enable
