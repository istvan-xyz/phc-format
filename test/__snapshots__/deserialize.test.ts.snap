// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`should thow errors if trying to deserialize an invalid phc string 1`] = `"phcString must be a non-empty string"`;

exports[`should thow errors if trying to deserialize an invalid phc string 2`] = `"phcString must contain a $ as first char"`;

exports[`should thow errors if trying to deserialize an invalid phc string 3`] = `"phcString contains too many fields: 5/4"`;

exports[`should thow errors if trying to deserialize an invalid phc string 4`] = `"phcString must contain a $ as first char"`;

exports[`should thow errors if trying to deserialize an invalid phc string 5`] = `"id must satisfy /^[a-z0-9-]{1,32}$/"`;

exports[`should thow errors if trying to deserialize an invalid phc string 6`] = `"params names must satisfy /^[a-z0-9-]{1,32}$/"`;

exports[`should thow errors if trying to deserialize an invalid phc string 7`] = `"params values must satisfy /^[a-zA-Z0-9/+.-]+$/"`;

exports[`should thow errors if trying to deserialize an invalid phc string 8`] = `"params must be in the format name=value"`;

exports[`should thow errors if trying to deserialize an invalid phc string 9`] = `"phcString contains unrecognized fields: unrecognized"`;

exports[`should thow errors if trying to deserialize an invalid phc string 10`] = `"phcString contains too many fields: 5/4"`;

exports[`should thow errors if trying to deserialize an invalid phc string 11`] = `"phcString contains unrecognized fields: unrecognized"`;
